### ISSUE TITLE
Checkboxes feels very laggy on iOS8 and somewhat laggy on desktop

### ISSUE THREAD

**Issue by rj33** [google/material-design-lite#1723]
At first I thought this was due to the 300ms delay issue, and it still might be , although Chrome is showing the same lag for me as Safari, and I thought the issue was resolved there, so I think it is not the 300ms touch->mouse event delay, but rather the animation transition being used.

Looking at the sass file it looks like a 280ms delay is being used on the animation, and that the visible effect of the animation isn't very visible until a long way into the animation. The ripple response is immediate, which is why I think the issue is more about the animation of the checkbox transition than the event processing, although I'm not 100% sure, as the desktop seems a lot less laggy than mobile, but there is still a somewhat annoying delay on desktop too.

I compared to the polymer checkbox which has the same look, and it is has a much better feel. It seems they are using a 140ms delay, but I suspect they might have a different easing which pushes most of the visible change up front thus giving the user a more responsive feel to their interaction.

I think radio buttons appear to have the same issue. 

Toggles feel ok, because while there is probably the same delay, the movement animation gives you immediate feedback that your interaction has been correctly received. BTW is it worth raising a feature request to have the toggles react to dragging the toggle handle, which seems to be a common interaction on these types of elements, and something that doesn't work on the current implementation - as a point of reference, the Polymer toggle button does react this way, so having two MD implementations with a very similar visual look, but differing behaviour is probably not ideal (especially if they both become as popular as they deserve to be!).

So in summary, would it be possible to
a) Reduce the animation time from 280ms to 140ms, and
b) Change the animation so that the change is much more visible earlier in the animation.

**Comment by surma** [comment]
Hey @rj33, I can’t quite follow which lag you are talking about. 

The ripple appears immediately on desktop for me and I can’t find any delay in our code. 
![checkbox](https://cloud.githubusercontent.com/assets/234957/10579158/724c517e-766e-11e5-8348-366f830d36a7.gif)

**Comment by rj33** [comment]
"delay" is probably the wrong word for it. In _checkbox.scss you have this:

 @include material-animation-default(0.28s);

Which looks like it is asking for a 280ms transition of the tick visibility transition. So while the tick starts to become visible immediacy, but presumably due to easing, it is not clearly 'there' until much later in the 280ms. When I look at your animation I see a delay between the initial circular click animation starting and the tick becoming clearly visible. For me at least this makes the checkbox response feel lagged.  While in theory the delays should be the same on mobile, it feels more lagged on an iphone 6 running iOS8 than it does on my desktop, I'm not sure why this is, but it feels like tick becomes visible much later in the transition on my device. If you compare the polymer case they use half the transition time, and may also have a different easing strategy (I didn't check the latter), and their checkbox feels a lot more responsive than the MDL checkbox (for me).

**Comment by surma** [comment]
Thanks for clarifying, @rj33!

What you are seeing on iOS is this (from the iOS Simulator):
![ioscheckbox](https://cloud.githubusercontent.com/assets/234957/10605719/8ec921a8-7725-11e5-8999-104d54d8c461.gif)

That is due too the 300ms tap delay that Safari still has. It has been removed [just recently](https://trac.webkit.org/changeset/191072) for non-zoomable pages. Nothing we can do about it, sadly :confused: 

The animation in my first post is exactly as intended, so desktop is fine.

**Comment by Garbee** [comment]
Note: Touch delay is related to issue #893. Nothing we can feasibly do on iOS to remove it where it is present. Tapping *slightly* longer than normal will negate the delay automatically, but how many users are going to know that?

**Comment by rj33** [comment]
Thanks @surma , that explains why the delay was worse on iOS. I'd mistakenly thought the 300ms delay was only applied to click events, which don't seem to be at play here, but it seems other events are also being impacted, so thanks for the heads up.

Personally I find the animation delay on desktop shown in your first animation annoying, as it makes the checkbox feel slightly unresponsive IMO, and is longer than other MD implementations use, but it seems fairly easy to override, so if nobody else has a problem with the desktop delay, happy to see this closed.

**Comment by rj33** [comment]
@garbee In theory the general case of 300ms delay should be removable by simply using touch events instead of mouse events on relevant devices, but I understand that may not be reasonably feasible within the current framework. If only Apple would take the pointer events work on board, this would all be very easy in the future given everyone else has committed to them now.
